test/mpmc_queue.ml

```diff
@@ -0,0 +1,186 @@
+  else if i >= Array.length t.buffer
+  then false
+  else begin
+    let cell = Array.unsafe_get t.buffer i in
+    match Atomic.get cell with
+    | Empty ->
+        if Atomic.compare_and_set cell Empty (Value x)
+        then true
+        else begin match Atomic.get cell with
+          | Tombstone -> push_s t x
+          | (Await fn) as old ->
+              assert (Atomic.compare_and_set cell old Tombstone) ;
+              fn x ;
+              true
+          | _ -> assert false
+        end
+    | (Await fn) as old ->
+        assert (Atomic.compare_and_set cell old Tombstone) ;
+        fn x ;
+        true
+    | Tombstone ->
+        push_s t x
+    | Value _ -> assert false
+  end
+
+let rec push t x =
+  let last = Atomic.get t.last in
+  if push_s last x
+  then ()
+  else begin
+    let rest = force_rest last in
+    let _ : bool = Atomic.compare_and_set t.last last rest in
+    push t x
+  end
+
+
+type 'a pop_result =
+  | Is_empty
+  | Wait_for_it
+  | Pop of 'a
+
+let rec pop_s t =
+  let current_head = Atomic.get t.head in
+  if current_head >= Array.length t.buffer
+  then Is_empty
+  else if current_head >= Atomic.get t.tail
+  then Wait_for_it
+  else
+    let i = Atomic.fetch_and_add t.head 1 in
+    if i >= Array.length t.buffer
+    then Is_empty
+    else
+      let cell = Array.unsafe_get t.buffer i in
+      if i >= Atomic.get t.tail
+      then tombstone t cell
+      else spinlock ~retry:spinlock_iterations t cell
+
+and tombstone t cell =
+  if Atomic.compare_and_set cell Empty Tombstone
+  then pop_s t
+  else begin match Atomic.get cell with
+       | (Value v) as old ->
+           assert (Atomic.compare_and_set cell old Tombstone) ;
+           Pop v
+       | _ -> assert false
+       end
+
+and spinlock ~retry t cell =
+  match Atomic.get cell with
+  | (Value v) as old ->
+      assert (Atomic.compare_and_set cell old Tombstone) ;
+      Pop v
+  | Empty when retry <= 0 ->
+      tombstone t cell
+  | Empty ->
+      Domain.cpu_relax () ;
+      spinlock ~retry:(retry - 1) t cell
+  | Tombstone | Await _ ->
+      assert false
+
+let rec pop t =
+  let first = Atomic.get t.first in
+  match pop_s first with
+  | Pop v -> Some v
+  | Wait_for_it -> None
+  | Is_empty ->
+      begin match Atomic.get first.rest with
+      | None -> None
+      | Some rest ->
+          let _ : bool = Atomic.compare_and_set t.first first rest in
+          pop t
+      end
+
+
+let await_pop_s ~pool t =
+  let i = Atomic.fetch_and_add t.head 1 in
+  if i >= Array.length t.buffer
+  then None
+  else
+    let cell = Array.unsafe_get t.buffer i in
+    match Atomic.get cell with
+    | (Value v) as old ->
+        assert (Atomic.compare_and_set cell old Tombstone) ;
+        Some v
+    | Empty ->
+        let promise, set_promise = Domainslib.Task.promise () in
+        if Atomic.compare_and_set cell Empty (Await set_promise)
+        then Some (Domainslib.Task.await pool promise)
```

Comment:
The three lines above are the fancy bit! When `pop` fails to obtain a value, it pushes a function onto the queue that will be called by the matching `push` to unlock the task :)